client/extensions/woocommerce/app/products/product-form-images.js

```diff
@@ -0,0 +1,164 @@
+/**
+ * External dependencies
+ */
+import React, { Component, PropTypes } from 'react';
+import classNames from 'classnames';
+import Gridicon from 'gridicons';
+import { localize } from 'i18n-calypso';
+
+/**
+ * Internal dependencies
+ */
+import Button from 'components/button';
+import FormLabel from 'components/forms/form-label';
+import FormSettingExplanation from 'components/forms/form-setting-explanation';
+import ImagePreloader from 'components/image-preloader';
+import ProductImageUploader from 'woocommerce/components/product-image-uploader';
+import Spinner from 'components/spinner';
+
+class ProductFormImages extends Component {
+	static propTypes = {
+		product: PropTypes.shape( {
+			images: PropTypes.array,
+		} ),
+		editProduct: PropTypes.func.isRequired,
+	};
+
+	constructor( props ) {
+		super( props );
+		const { product } = this.props;
+		const images = product.images || [];
+		this.state = {
+			images,
+		};
+	}
+
+	onUpload = ( file ) => {
+		const { product, editProduct } = this.props;
+		const images = product.images && [ ...product.images ] || [];
+		images.push( {
+			id: file.ID,
```

Comment:
I see three slightly different "shapes" for an image object. Just id+src here, or spreaded attributes, id and src below, or placeholder, transient, id and src below that. This might make it hard to work on this code in the future. Would having a small helper function that builds an image object help avoid that?